### DISCUSSION TITLE
Installing additional R packages in an environment does not work (reliably)

### DISCUSSION THREAD

**Discussion by kohlman** [spack/spack#27795] [Q&A] [answered]
Dear all,

I am instructing people to use Spack on a cluster to run analyses and it is an amazing tool (but I am not an IT person per se). I would like to use environments for this purpose because it is really elegant, but I do have a certain issue that I cannot solve:

In many cases, especially for R and Python, I need a package that is not already included in the list that can be installed via Spack (like r-tidyverse or any of the r-... packages). This would just be one of the many packages that people haven't prepared yet as a separate Spack package. I would then have to install packages within R or with pip for Python etc.

This seems to work for a short time, but at some point the packages are then not available anymore within the environment. They seem to just disappear, e.g., after reactivating the environment later. I have looked at the locations where the packages are stored and my hunch is that somehow the environment R/Python installations get overwritten by the base installations of those programs. It definitely seems to happen after another concretization/installation when the view is re-adapted.

Well, one workaround is that I do these additional installations in the base environment by loading R/Python there first and then installing the required packages. But this can get complicated and I also feel that the advantages of the environment approach are thwarted in this case.

I feel like I am missing a mechanism or command that would solve this. I thought I might have to change some view options to avoid the overwriting of the view by the base environment.

Sorry if I didn't find the same or similar discussions in the forum!

Thanks for any hints and best,

Axel

**Comment by glennpj** [comment, marked as answer]
Axel,

This is a permissions issue at the OS level so I am not sure there is a way to solve it with spack. This is not limited to spack environments although the manifestation of issues would look a bit different. When you install packages with spack, you obviously have write access to the installation directories. If you initiate a Python or R session then those programs will consider the respective install directories to be the main package/library path. If you then install packages it will write to those directories. For non-spack environments, the Python and R directories would be used. However, if r- or py- programs are also loaded then those directories may be considered the location to place new packages. For spack environments using a view, the view directory would be the directory those package installs would write to. The extra twist with environment views is that they get regenerated.

You will have to be careful about where you are installing non-spack packages to so they do not get mixed with spack packages. For Python, you could use the `--user` flag to `pip` which will install things in `~/.local`. For R, you could use a private R library, usually under your $HOME directory. I generally recommend using virtual environments for Python and packrat/renv for R to isolate the environments completely. Those options come with their own challenges however.

Even if using separate paths and/or virtual environments it is still possible to forget and wind up mixing spack and non-spack packages. One option might be to write a wrapper script around spack that would toggle write access to specific directories, like the R and Python directories, or in the case of environments, the view directory. At the beginning of the wrapper, write permission is toggled on, and toggled off at the end. That will prevent Python and R from being able to write to the directory that it initially wants to.

I hope that helps.

**Reply by kohlman** [reply]
Glenn,

Thanks so much for the input. In a sense, I am glad that this is a generic problem and I wasn't just too dumb to find the answer. I was hoping that there would be some kind solution because it must be such a common issue. Working with R or Python, in most cases you would need an extra package. But you confirmed my suspicion that the views regenerate and that you therefore lose the manually installed packages.

Thanks for the tips on completely encapsulating environments! I will give it a try.

Just as a side note: My own current solution is that I install all the necessary stuff in the base environment and have a load script for the relevant packages at startup.

Best,

Axel
